test/e2e: add deploying separate OAuth server to util

stlaz opened the pull request:
This PR makes it possible to allow deploying a separate OAuth server for e2e tests

/assign @sallyom 

It ss a part of https://github.com/openshift/origin/issues/23258

sallyom commented on test/extended/testdata/oauthserver/session-secret.yaml:
```diff
@@ -0,0 +1,8 @@
+metadata:
+  labels:
+    app: test-oauth-server
+  name: session # this secret should be generated
```
yup, i know, i copied it from my cluster to get it working lol

sallyom replied:
generated now.

stlaz commented on test/extended/oauth/sampleoauthserver.go:
```diff
@@ -0,0 +1,93 @@
+			UseAsChallenger: true,
+			UseAsLogin:      true,
+			MappingMethod:   "claim",
+			Provider:        runtime.RawExtension{Raw: encodeOrDie(&osinv1.HTPasswdPasswordIdentityProvider{File: exutil.GetPathFromConfigMapSecretName("htpasswd-secret", "htpasswd")})},
```
The encode is perhaps still a bit too uncomfortable...

stlaz commented:
TODO: still needs to config SNI for the OAuth so that we don't get bad issuer errors...

stlaz commented on test/extended/util/oauthserver.go:
```diff
@@ -0,0 +1,594 @@
+}
+
+func RequestTokenForUser(reqOpts *tokencmd.RequestTokenOptions, username, password string) (string, error) {
+	reqOptsCopy := *reqOpts
```
This is probably wrong and maybe a bit unnecessary...

stlaz commented:
Removed FIXME and hopefully fixed test/verify

sallyom commented on test/extended/util/oauthserver.go:
```diff
@@ -0,0 +1,590 @@
+	// the oauth server needs access to kube-system configmaps/extension-apiserver-authentication
+	oauthSARolebinding := &rbacv1.ClusterRoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: oc.Namespace(), // TODO: probably something more cleaver?
```
this is fine i think

stlaz replied:
Ok, I'll remove the comment then. What a typo tho.

stlaz commented:
I added the cleanup changes from https://github.com/stlaz/origin/pull/3 and added some more utility code for the tests.

This can now be reviewed.

stlaz commented:
/test cmd
timeouts

deads2k commented on test/extended/util/oauthserver/oauthserver.go:
```diff
@@ -0,0 +1,546 @@
+// DeployOAuthServer - deployes an instance of an OpenShift OAuth server
+// very simplified for now
+// returns OAuth server url, cleanup function, error
+func DeployOAuthServer(oc *exutil.CLI, idps []osinv1.IdentityProvider, configMaps []corev1.ConfigMap, secrets []corev1.Secret) (*tokencmd.RequestTokenOptions, func(), error) {
```
if you end up having to expand this,  would consider creating chainable functions just for producing a configuration.  This is ok to start, but think `DefaultConfig().WithIDP().SetMyFunc().ToConfigOrDie()`

deads2k replied:
> prefer your own struct with `NewFromCLI` constructor. Consider what would happen if you were disconnected from kube ginkgo.

This doesn't haven't to happen before merge.

deads2k commented on test/extended/util/oauthserver/oauthserver.go:
```diff
@@ -0,0 +1,546 @@
+// DeployOAuthServer - deployes an instance of an OpenShift OAuth server
+// very simplified for now
+// returns OAuth server url, cleanup function, error
+func DeployOAuthServer(oc *exutil.CLI, idps []osinv1.IdentityProvider, configMaps []corev1.ConfigMap, secrets []corev1.Secret) (*tokencmd.RequestTokenOptions, func(), error) {
```
prefer your own struct with `NewFromCLI` constructor.  Consider what would happen if you were disconnected from kube ginkgo.

deads2k commented on test/extended/util/oauthserver/oauthserver.go:
```diff
@@ -0,0 +1,546 @@
+	}
+
+	// get the route of the future OAuth server
+	route, err := oc.AdminRouteClient().RouteV1().Routes(oc.Namespace()).Get(RouteName, metav1.GetOptions{})
```
Where did this get created?

sallyom replied:
it's in test/extended/testdata/oauth-server.yaml

stlaz replied:
The file is now `test/extended/testdata/oauthserver/oauth-network.yaml` and I added a comment about what it's creating.

deads2k commented on test/extended/util/oauthserver/oauthserver.go:
```diff
@@ -0,0 +1,546 @@
+		return nil, cleanups, fmt.Errorf("error encoding the OSIN config")
+	}
+
+	if err = oc.AsAdmin().Run("create").Args("configmap", "oauth-config", "--from-literal", fmt.Sprintf("oauth.conf=%s", string(configBytes))).Execute(); err != nil {
```
create inline please

deads2k commented on test/extended/util/oauthserver/oauthserver.go:
```diff
@@ -0,0 +1,546 @@
+			if err != nil {
+				return false, fmt.Errorf("could not retrieve the browser client: %v", err)
+			}
+			if reflect.DeepEqual(challengeClient.RedirectURIs, challengeRedirectURIsBackup.RedirectURIs) ||
```
This is weird.  Is there some auto-update happening?

stlaz commented:
With https://github.com/openshift/origin/pull/23259 merged, I was able to remove the oauthclients part, also addressed the comments required for this to merge.

stlaz commented:
/test e2e-aws-upgrade

stlaz commented:
/test e2e-aws
The failure is in the actual test, I do not want to introduce more flakes, but I would like to check how reproducible it is

stlaz commented:
The above issue does not seem to be reproducible (I tried to rerun this many times manually), this should be ok to merge.

sallyom commented:
/lgtm

deads2k commented:
/approve

smarterclayton commented on test/extended/oauth/htpasswd.go:
```diff
@@ -0,0 +1,80 @@
+			MappingMethod:   "claim",
+			Provider:        *htpasswdProvider,
+		}
+		tokenReqOpts, cleanup, err := utiloauth.DeployOAuthServer(oc, []osinv1.IdentityProvider{htpasswdConfig}, nil, secrets)
```
Any test that changes the configuration of the cluster must have `[Disruptive]` or `[Serial]` depending on which suite.  Since this changes core configuration, you may not be part of the parallel suite.

stlaz replied:
This deploys a standalone OAuth server and does not change the cluster config, only the configuration of the specific OAuth server that’s being deployed.

smarterclayton commented:
/hold

stlaz commented:
Rebased on current master.

@smarterclayton could you please be specific about which cluster configuration change worries you? The OAuth server is configured against the API server but by itself, it does not alter any cluster configuration. It will, of course, add users and identities which need to be cleaned up after running such a test, and it does add cluster-admin role to the SA running the OAuth server. Is that the problem?

deads2k commented:
@smarterclayton different oauth server.  this one is intentionally independent and unrelated to the one for the cluster

/hold cancel
/lgtm

deads2k commented:
> @smarterclayton could you please be specific about which cluster configuration change worries you? The OAuth server is configured against the API server but by itself, it does not alter any cluster configuration. It will, of course, add users and identities which need to be cleaned up after running such a test, and it does add cluster-admin role to the SA running the OAuth server. Is that the problem?

You can add wide permission and create unique cluster scoped resources as long as you clean them up properly.  See `exutil.CLI.AddResourceToDelete` and it's many usages.

stlaz commented:
/retest

stlaz commented:
/retest

stlaz commented:
/retest
fails to pull images apparently

deads2k commented:
/lgtm

stlaz commented:
/hold
The oauth-server image that we retrieve from inside cluster changed and no longer contains hypershift binary, no reason to torture CI until I fix that.

stlaz commented:
/hold cancel
updated the image, but I noticed the test were failing more often now so I added 5 seconds initial delay to readiness probe and the tests were passing 100% of the time for me
@deads2k see if -^ is ok, please

stlaz commented:
/retest

stlaz commented:
/test cmd

sttts commented on test/extended/testdata/oauthserver/oauth-pod.yaml:
```diff
@@ -0,0 +1,92 @@
+        path: /healthz
+        port: 6443
+        scheme: HTTPS
+      initialDelaySeconds: 5
```
which means it succeeds with 35s, but flakes with 35? Looks fragile.

stlaz replied:
I checked the logs, the OAuth server was actually alive but never got the request, something else seems to be wrong here

deads2k commented:
/lgtm
